lightning/restore/restore.go

```diff
@@ -304,48 +305,207 @@ outside:
+				if sql != "" {
+					stmts, err := createTableIfNotExistsStmt(worker.glue.GetParser(), sql, dbMeta.Name, viewMeta.Name)
+					if err != nil {
+						worker.throw(err)
+						return
+					}
+					for _, sql := range stmts {
+						restoreSchemaJob.stmts = append(restoreSchemaJob.stmts, &schemaStmt{
+							tblName:  viewMeta.Name,
+							stmtType: schemaCreateView,
+							sql:      sql,
+						})
+					}
+				}
 			}
+			worker.wg.Add(1)
+			worker.jobCh <- restoreSchemaJob
+			// we don't support restore views concurrency, cauz it maybe will raise a error
+			worker.wg.Wait()
 		}
+		worker.quit()
+	}
+}
 
-		// restore views. Since views can cross database we must restore views after all table schemas are restored.
-		for _, dbMeta := range rc.dbMetas {
-			if len(dbMeta.Views) > 0 {
-				task := log.With(zap.String("db", dbMeta.Name)).Begin(zap.InfoLevel, "restore view schema")
-				viewsSchema := make(map[string]string)
-				for _, viewMeta := range dbMeta.Views {
-					viewsSchema[viewMeta.Name] = viewMeta.GetSchema(ctx, rc.store)
+func (worker *restoreSchemaWorker) doJob() {
+	var logger log.Logger
+	var session checkpoints.Session
+	defer func() {
+		if session != nil {
+			session.Close()
+		}
+	}()
+	for {
+		select {
+		case <-worker.ctx.Done():
+			return
+		case job := <-worker.jobCh:
+			if job == nil {
+				return
+			}
+			var err error
+			if session == nil {
+				session, err = worker.glue.GetSession(worker.ctx)
+				if err != nil {
+					worker.throw(err)
+					return
 				}
-				err := InitSchema(ctx, rc.tidbGlue, dbMeta.Name, viewsSchema)
-
+			}
+			for _, stmt := range job.stmts {
+				if stmt.stmtType == schemaCreateDatabase {
+					logger = log.With(zap.String("db", job.dbName))
+				} else if stmt.stmtType == schemaCreateTable {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+				} else if stmt.stmtType == schemaCreateView {
+					logger = log.With(zap.String("table", common.UniqueTable(job.dbName, stmt.tblName)))
+				}
+				task := logger.Begin(zap.DebugLevel, fmt.Sprintf("execute SQL: %s", stmt.sql))
+				_, err = session.Execute(worker.ctx, stmt.sql)
 				task.End(zap.ErrorLevel, err)
 				if err != nil {
-					return errors.Annotatef(err, "restore view schema %s failed", dbMeta.Name)
+					switch stmt.stmtType {
+					case schemaCreateDatabase:
+						err = errors.Annotatef(err, "restore database schema %s failed", job.dbName)
+					case schemaCreateTable:
+						err = errors.Annotatef(err, "restore table schema %s failed", stmt.tblName)
+					case schemaCreateView:
+						err = errors.Annotatef(err, "restore view schema %s failed", stmt.tblName)
+					}
+					worker.wg.Done()
+					worker.throw(err)
+					return
 				}
 			}
+			worker.wg.Done()
+		}
+	}
+}
 
+func (worker *restoreSchemaWorker) wait() error {
+	defer worker.quit()
+	select {
+	case err := <-worker.errCh:
+		return err
+	case <-worker.ctx.Done():
```

Comment:
This is a problem here, if this Done event is produced by the worker itself, we can return nil here. But if the Done event is caused by parent context, return `ctx.Err()` is better. So It's better not depend on context.Quit() to finish sub routines here

Reply:
I will replace `worker.quit()` to `close(worker.jobCh)`